src/libmongoc/tests/unified/test-diagnostics.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2020-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "test-diagnostics.h"
+
+#include "utlist.h"
+#include "TestSuite.h"
+#include "common-thread-private.h"
+#include <signal.h>
+
+typedef struct _msg_t {
+   char *string;
+   struct _msg_t *next;
+} msg_t;
+
+typedef struct {
+   msg_t *test_info;
+   msg_t *error_info;
+   bson_mutex_t mutex;
```

Comment:
It looks like this mutex is only used when adding to this struct (in `_test_diagnostics_add`). Should we also use it before reading from this struct? We do in other places in our codebase. For example:

```
mongoc_topology_description_type_t
_mongoc_topology_get_type (mongoc_topology_t *topology)
{
   mongoc_topology_description_type_t td_type;

   bson_mutex_lock (&topology->mutex);

   td_type = topology->description.type;

   bson_mutex_unlock (&topology->mutex);

   return td_type;
}
```
It seems like we could run into the case where information is changing as we're reading it. Though, I'm not entirely sure.